{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explore the Data\n",
    "My own working notes to better understand what data I have access to that may be helpful in relating HIV-Associated Neurocognitive Disorder (HAND) with HIV sequence data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Outline\n",
    "1. [Overview](#overview)  \n",
    "2. [Sequence Data](#seqs)  \n",
    "3. [Meta Data](#meta) \n",
    "4. [Preliminary Questions](#questions)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='overview'></a>\n",
    "## 1. Overview\n",
    "\n",
    "### Ogishi and Yotsuyanagi \n",
    "Paper: https://retrovirology.biomedcentral.com/articles/10.1186/s12977-018-0401-x  \n",
    "Data: https://github.com/masato-ogishi/HANDPrediction/tree/master/inst\n",
    "\n",
    "Three data sets:\n",
    "- HAND Database (data from the HAND Database)\n",
    "- LANL (data from the Los Alamos National Laboratory HIV Database)\n",
    "- Metadataset (data from the authors own curated set of sequences and metadata\n",
    "\n",
    "Three types of files:\n",
    "- __Sequence File__ C2V3C3_{DataSetName}_AA.fasta (amino acid sequence alignments for the C2V3C3 region of the HIV Envelope gene in fasta format)\n",
    "- __Metadata File__ {DataSetName}.csv (metadata associated with each sequence {i.e. Hand status, patient id, accession number, publication, etc})\n",
    "- __Sample Tissue File__ {DataSetName}_DesignSheet.csv (A mapping of Sample.Tissue { from the other .csv file} to Sample.Tissue.Category)\n",
    "\n",
    "### Holman and Gabuzda\n",
    "Paper: https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0049538  \n",
    "Data: \n",
    "- (Validation seqs): https://github.com/veg/idepi/tree/master/examples/dementia/validation\n",
    "- (MetaData): https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0049538#s5\n",
    "\n",
    "A validation dataset that contains sequences from 10 individuals diagnosed with HAD, we have the sequences in the idepi codebase; just one file (HAD.fas) that contains the nucleotide sequences in fasta format.\n",
    "\n",
    "Metadata for the training and testing datasets were available as excel files in the suplemental material of the paper. These are the two csv files saved in the HolmanAndGabuzda directory.\n",
    "\n",
    "### idepi/Hepler\n",
    "Paper: https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003842  \n",
    "Data: https://github.com/veg/idepi/tree/master/examples/dementia\n",
    "\n",
    "This is esentially the dataset used by Holman and Gabuzda (see above). The idepi project (by Hepler, et al.) used the Holman and Gabuzda data set for testing.\n",
    "\n",
    "- The Training dataset: \"HIVBrainSeqDBHepler.fasta\" and \"HAD.csv\" (which just contains the HAD status)\n",
    "- The Testing/validation dataset: \"HAD.fas\" (The fasta file for the 10 patients, all of whom were diagnosed with HAD\n",
    "\n",
    "### HIVBrainSeqDB\n",
    "Paper: https://aidsrestherapy.biomedcentral.com/articles/10.1186/1742-6405-7-43  \n",
    "Data: The website has moved here http://hivbrainseqdb.dfci.harvard.edu/HIVSeqDB/; (http://www.HIVBrainSeqDB.org no longer works). \n",
    "\n",
    "I downloaded the entire HIV BrainSequence Database from http://hivbrainseqdb.dfci.harvard.edu/HIVSeqDB/ on 10-10-2018 and saved it in the data folder.\n",
    "\n",
    "One data set with two files:\n",
    "- HIVSeqDB.fasta (The nucleotide sequences in fasta format, from the HIV Brain Sequence Database)\n",
    "- HIVSeqDB-MetaData.txv (the Metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='seqs'></a>\n",
    "## 2. Sequence Data (Fasta Files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in the fasta files and create a list of ids and a list of sequences.\n",
    "Files = {}\n",
    "Files['C2V3C3_HANDDatabase_AA.fasta'] = {'filePath': './OgishiAndYotsuyanagi/C2V3C3_HANDDatabase_AA.fasta'}\n",
    "Files['C2V3C3_LANL_AA.fasta'] = {'filePath': './OgishiAndYotsuyanagi/C2V3C3_LANL_AA.fasta'}\n",
    "Files['C2V3C3_Metadataset_AA.fasta'] = {'filePath': './OgishiAndYotsuyanagi/C2V3C3_Metadataset_AA.fasta'}\n",
    "Files['HIVBrainSeqDB.fasta'] = {'filePath': './HIVBrainSeqDB/HIVSeqDB.fasta'}\n",
    "Files['HIVBrainSeqDBHepler.fasta'] = {'filePath': './idepi/HIVBrainSeqDBHepler.fasta'}\n",
    "Files['HAD.fas'] = {'filePath': './idepi/HAD.fas'}\n",
    "\n",
    "for file in Files:\n",
    "    with open(Files[file][\"filePath\"]) as fasta_file:\n",
    "        ids = []\n",
    "        sequences = []\n",
    "        for title, sequence in SimpleFastaParser(fasta_file):\n",
    "            ids.append(title.split(None, 1)[0])  # First word is ID\n",
    "            sequences.append(sequence)\n",
    "        Files[file][\"ids\"] = ids\n",
    "        Files[file][\"sequences\"] = sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C2V3C3_HANDDatabase_AA.fasta\n",
      "number of sites: {189}\n",
      "number of sequences: 1690\n",
      "\n",
      "\n",
      "C2V3C3_LANL_AA.fasta\n",
      "number of sites: {189}\n",
      "number of sequences: 21609\n",
      "\n",
      "\n",
      "C2V3C3_Metadataset_AA.fasta\n",
      "number of sites: {189}\n",
      "number of sequences: 2494\n",
      "\n",
      "\n",
      "HIVBrainSeqDB.fasta\n",
      "number of sites: {2560, 2562, 1541, 1030, 2565, 1032, 1033, 1545, 2571, 1020, 1544, 2574, 1039, 1036, 2577, 1042, 1044, 2583, 1048, 1047, 2586, 1023, 2589, 2592, 2595, 2601, 1066, 2607, 2608, 2610, 2613, 2614, 2616, 2617, 1081, 1082, 2625, 1093, 2119, 2634, 2635, 592, 2641, 595, 2644, 598, 2647, 601, 2650, 604, 607, 610, 613, 2665, 106, 619, 105, 622, 1650, 628, 631, 1153, 1164, 1167, 1170, 1173, 1179, 1182, 160, 161, 162, 163, 164, 1185, 1188, 170, 1197, 176, 1200, 1203, 181, 182, 1206, 2229, 1209, 1212, 189, 1215, 1218, 195, 1260, 198, 205, 206, 207, 1230, 1233, 210, 211, 1236, 213, 215, 216, 1239, 218, 219, 1242, 1245, 222, 1248, 225, 226, 1266, 1251, 230, 231, 1254, 1257, 234, 235, 236, 237, 238, 1263, 240, 241, 242, 243, 244, 245, 246, 1269, 1272, 249, 250, 251, 252, 253, 254, 255, 256, 257, 258, 259, 1281, 261, 262, 263, 264, 265, 266, 1287, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 278, 1275, 1278, 294, 297, 298, 300, 301, 303, 304, 310, 313, 316, 1284, 319, 322, 325, 328, 1290, 2422, 2425, 2434, 2439, 2446, 2458, 2464, 2467, 2468, 2470, 429, 430, 2482, 948, 2490, 2491, 1986, 2523, 2526, 2529, 2530, 2532, 998, 2535, 1001, 1002, 9706, 2541, 9711, 2547, 2550, 1016, 2553, 1018, 2555, 2556, 1017, 2559}\n",
      "number of sequences: 2552\n",
      "\n",
      "\n",
      "HIVBrainSeqDBHepler.fasta\n",
      "number of sites: {2563, 1027, 1029, 1034, 1035, 2575, 2576, 1021, 2578, 2581, 2584, 1053, 2593, 2602, 2603, 1068, 1069, 2614, 1080, 585, 591, 594, 104, 1150, 1153, 1156, 1165, 1168, 1171, 2202, 158, 159, 160, 161, 1183, 1188, 1191, 1194, 1197, 1200, 1203, 1215, 1218, 196, 1221, 1224, 205, 1230, 211, 213, 214, 1239, 217, 1245, 223, 228, 229, 1257, 233, 234, 235, 237, 238, 1263, 239, 1265, 240, 243, 1268, 246, 1271, 263, 267, 270, 271, 273, 274, 294, 297, 300, 2392, 2395, 2416, 9585, 9590, 2428, 2437, 2438, 2440, 2460, 424, 937, 1962, 425, 2492, 2498, 2499, 2501, 2504, 2510, 2516, 2519, 986, 2524, 2525, 990, 989, 2528, 2530, 2539, 1004, 1005, 1006, 2542, 1008, 2545, 1011, 1525, 1526, 2551, 1018, 1020, 2557}\n",
      "number of sequences: 1085\n",
      "\n",
      "\n",
      "HAD.fas\n",
      "number of sites: {2688, 2529, 2689, 2532, 2535, 2695, 2697, 2683, 2699, 2700, 2541, 2574, 429, 2701, 2704, 2670, 2263, 2713, 2523}\n",
      "number of sequences: 75\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Look at how many sequences and sites.\n",
    "for file in Files:\n",
    "    print(file)\n",
    "    sequences = Files[file][\"sequences\"]\n",
    "    seqLengths = []\n",
    "    for seq in sequences:\n",
    "        seqLengths.append(len(seq))\n",
    "    \n",
    "    uniqueSeqLengths = set(seqLengths)\n",
    "    print('number of sites: ' + str(uniqueSeqLengths))\n",
    "    print('number of sequences: ' + str(len(seqLengths)))\n",
    "    print('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='meta'></a>\n",
    "## 3. Metadata (csv Files)\n",
    "We won't print the HANDDatabase or LANL files from the Ogishi and Yotsuyanagi paper; they are very similar to the two files printed below for the Metadataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5386, 20)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Clinical.Status</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Reference_PublicationYear</th>\n",
       "      <th>Reference_FirstAuthor</th>\n",
       "      <th>Reference_PMID</th>\n",
       "      <th>LANL_Pub.id</th>\n",
       "      <th>LANL_SE.id</th>\n",
       "      <th>LANL_Patient.Id</th>\n",
       "      <th>Patient.Sex</th>\n",
       "      <th>Number.of.patient.seqs</th>\n",
       "      <th>Georegion</th>\n",
       "      <th>Coreceptor</th>\n",
       "      <th>Sample.Tissue</th>\n",
       "      <th>Culture.Method</th>\n",
       "      <th>Viral.load</th>\n",
       "      <th>CD4.count</th>\n",
       "      <th>CD8.count</th>\n",
       "      <th>Days.from.Infection</th>\n",
       "      <th>Days.from.Seroconversion</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HAND</td>\n",
       "      <td>AF125810</td>\n",
       "      <td>1999</td>\n",
       "      <td>Shapshak</td>\n",
       "      <td>10381169</td>\n",
       "      <td>9244</td>\n",
       "      <td>231359</td>\n",
       "      <td>885.0</td>\n",
       "      <td>F</td>\n",
       "      <td>52.0</td>\n",
       "      <td>North America</td>\n",
       "      <td>NaN</td>\n",
       "      <td>brain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late</td>\n",
       "      <td>ACCCCACTCTGTGTTACTTTAAATTGCACTGATGCTAATTGGAGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HAND</td>\n",
       "      <td>AF125811</td>\n",
       "      <td>1999</td>\n",
       "      <td>Shapshak</td>\n",
       "      <td>10381169</td>\n",
       "      <td>9244</td>\n",
       "      <td>231358</td>\n",
       "      <td>885.0</td>\n",
       "      <td>F</td>\n",
       "      <td>52.0</td>\n",
       "      <td>North America</td>\n",
       "      <td>NaN</td>\n",
       "      <td>brain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late</td>\n",
       "      <td>ACCCCACTCTGTGTTACTTTAAATTGCACTAATTGGGAGAATACTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HAND</td>\n",
       "      <td>AF125812</td>\n",
       "      <td>1999</td>\n",
       "      <td>Shapshak</td>\n",
       "      <td>10381169</td>\n",
       "      <td>9244</td>\n",
       "      <td>231357</td>\n",
       "      <td>885.0</td>\n",
       "      <td>F</td>\n",
       "      <td>52.0</td>\n",
       "      <td>North America</td>\n",
       "      <td>NaN</td>\n",
       "      <td>brain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late</td>\n",
       "      <td>ACCCCACTCTGTGTTACTTTAAATTGCACTAATTGGGAGAATATGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HAND</td>\n",
       "      <td>AF125813</td>\n",
       "      <td>1999</td>\n",
       "      <td>Shapshak</td>\n",
       "      <td>10381169</td>\n",
       "      <td>9244</td>\n",
       "      <td>231356</td>\n",
       "      <td>885.0</td>\n",
       "      <td>F</td>\n",
       "      <td>52.0</td>\n",
       "      <td>North America</td>\n",
       "      <td>NaN</td>\n",
       "      <td>brain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late</td>\n",
       "      <td>ACCCCACTCTGTGTTACTTTAAATTGCACTGATGCTAATTGGAGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HAND</td>\n",
       "      <td>AF125814</td>\n",
       "      <td>1999</td>\n",
       "      <td>Shapshak</td>\n",
       "      <td>10381169</td>\n",
       "      <td>9244</td>\n",
       "      <td>231355</td>\n",
       "      <td>885.0</td>\n",
       "      <td>F</td>\n",
       "      <td>52.0</td>\n",
       "      <td>North America</td>\n",
       "      <td>NaN</td>\n",
       "      <td>brain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late</td>\n",
       "      <td>ACCCCACTCTGTGTTACTTTAAATTGCACTGATTGGAAGAATAGTA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Clinical.Status Accession  Reference_PublicationYear Reference_FirstAuthor  \\\n",
       "0            HAND  AF125810                       1999              Shapshak   \n",
       "1            HAND  AF125811                       1999              Shapshak   \n",
       "2            HAND  AF125812                       1999              Shapshak   \n",
       "3            HAND  AF125813                       1999              Shapshak   \n",
       "4            HAND  AF125814                       1999              Shapshak   \n",
       "\n",
       "   Reference_PMID  LANL_Pub.id  LANL_SE.id  LANL_Patient.Id Patient.Sex  \\\n",
       "0        10381169         9244      231359            885.0           F   \n",
       "1        10381169         9244      231358            885.0           F   \n",
       "2        10381169         9244      231357            885.0           F   \n",
       "3        10381169         9244      231356            885.0           F   \n",
       "4        10381169         9244      231355            885.0           F   \n",
       "\n",
       "   Number.of.patient.seqs      Georegion Coreceptor Sample.Tissue  \\\n",
       "0                    52.0  North America        NaN         brain   \n",
       "1                    52.0  North America        NaN         brain   \n",
       "2                    52.0  North America        NaN         brain   \n",
       "3                    52.0  North America        NaN         brain   \n",
       "4                    52.0  North America        NaN         brain   \n",
       "\n",
       "  Culture.Method  Viral.load  CD4.count  CD8.count  Days.from.Infection  \\\n",
       "0            NaN         NaN        NaN        NaN                  NaN   \n",
       "1            NaN         NaN        NaN        NaN                  NaN   \n",
       "2            NaN         NaN        NaN        NaN                  NaN   \n",
       "3            NaN         NaN        NaN        NaN                  NaN   \n",
       "4            NaN         NaN        NaN        NaN                  NaN   \n",
       "\n",
       "  Days.from.Seroconversion                                           Sequence  \n",
       "0                     late  ACCCCACTCTGTGTTACTTTAAATTGCACTGATGCTAATTGGAGGA...  \n",
       "1                     late  ACCCCACTCTGTGTTACTTTAAATTGCACTAATTGGGAGAATACTG...  \n",
       "2                     late  ACCCCACTCTGTGTTACTTTAAATTGCACTAATTGGGAGAATATGT...  \n",
       "3                     late  ACCCCACTCTGTGTTACTTTAAATTGCACTGATGCTAATTGGAGGA...  \n",
       "4                     late  ACCCCACTCTGTGTTACTTTAAATTGCACTGATTGGAAGAATAGTA...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The Ogishi currated metadata dataset.\n",
    "metadata = pd.read_csv('./OgishiAndYotsuyanagi/Metadataset.csv')\n",
    "print (metadata.shape)\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(20, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample.Tissue</th>\n",
       "      <th>Sample.Tissue.Category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>blood</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>brain</td>\n",
       "      <td>CNS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>colon</td>\n",
       "      <td>Others</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CSF</td>\n",
       "      <td>CNS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>diaphragm</td>\n",
       "      <td>Others</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample.Tissue Sample.Tissue.Category\n",
       "0         blood                  Blood\n",
       "1         brain                    CNS\n",
       "2         colon                 Others\n",
       "3           CSF                    CNS\n",
       "4     diaphragm                 Others"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The \"DesignSheet\" for the Ogishi currated metadata dataset.\n",
    "metadataDesignSheet = pd.read_csv('./OgishiAndYotsuyanagi/Metadataset_SampleTissue_DesignSheet.csv')\n",
    "print(metadataDesignSheet.shape)\n",
    "metadataDesignSheet.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(861, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>HAD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DQ976387</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DQ976386</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DQ976385</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DQ976384</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DQ976383</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID  HAD\n",
       "0  DQ976387   50\n",
       "1  DQ976386   50\n",
       "2  DQ976385   50\n",
       "3  DQ976384   50\n",
       "4  DQ976383   50"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The HAD values from idepi associated with the Holman validation dataset.\n",
    "HAD = pd.read_csv('./idepi/HAD.csv')\n",
    "print(HAD.shape)\n",
    "HAD.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(78, 12)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Publication</th>\n",
       "      <th>Patient</th>\n",
       "      <th>Sequence Count</th>\n",
       "      <th>Neurocognitive Diagnosis</th>\n",
       "      <th>Disease Status</th>\n",
       "      <th>ART Treatment</th>\n",
       "      <th>CD4 Count</th>\n",
       "      <th>Viral Load (brain) copies/10e6 cells</th>\n",
       "      <th>Viral Load (plasma) copies/10e6 cells</th>\n",
       "      <th>Viral Load (lymphoid) copies/10e6 cells</th>\n",
       "      <th>Tissue Source</th>\n",
       "      <th>Year of Death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gatanaga, 1999 [1]</td>\n",
       "      <td>subject 4</td>\n",
       "      <td>6</td>\n",
       "      <td>HAD: severity not specified</td>\n",
       "      <td>AIDS (deceased)</td>\n",
       "      <td>AZT</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Lamers, 2009 [2]</td>\n",
       "      <td>DY</td>\n",
       "      <td>89</td>\n",
       "      <td>HAD: severe</td>\n",
       "      <td>AIDS (deceased)</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7000000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>University of California at San Francisco AIDS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Lamers, 2009 [2]</td>\n",
       "      <td>GA</td>\n",
       "      <td>52</td>\n",
       "      <td>HAD: severe</td>\n",
       "      <td>AIDS (deceased)</td>\n",
       "      <td>yes</td>\n",
       "      <td>400.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>University of California at San Francisco AIDS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Lamers, 2009 [2]</td>\n",
       "      <td>CX</td>\n",
       "      <td>116</td>\n",
       "      <td>HAD: severity not specified</td>\n",
       "      <td>AIDS (deceased)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>University of California at San Francisco AIDS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Li, 1991 [3]</td>\n",
       "      <td>YU</td>\n",
       "      <td>2</td>\n",
       "      <td>HAD: severe</td>\n",
       "      <td>AIDS (deceased)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Publication    Patient  Sequence Count     Neurocognitive Diagnosis  \\\n",
       "0  Gatanaga, 1999 [1]  subject 4               6  HAD: severity not specified   \n",
       "1    Lamers, 2009 [2]         DY              89                  HAD: severe   \n",
       "2    Lamers, 2009 [2]         GA              52                  HAD: severe   \n",
       "3    Lamers, 2009 [2]         CX             116  HAD: severity not specified   \n",
       "4        Li, 1991 [3]         YU               2                  HAD: severe   \n",
       "\n",
       "    Disease Status ART Treatment  CD4 Count  \\\n",
       "0  AIDS (deceased)           AZT        5.0   \n",
       "1  AIDS (deceased)           yes        NaN   \n",
       "2  AIDS (deceased)           yes      400.0   \n",
       "3  AIDS (deceased)       unknown        NaN   \n",
       "4  AIDS (deceased)           NaN        NaN   \n",
       "\n",
       "   Viral Load (brain) copies/10e6 cells  \\\n",
       "0                                   NaN   \n",
       "1                                   NaN   \n",
       "2                                   NaN   \n",
       "3                                   NaN   \n",
       "4                                   NaN   \n",
       "\n",
       "   Viral Load (plasma) copies/10e6 cells  \\\n",
       "0                                    NaN   \n",
       "1                              7000000.0   \n",
       "2                                    NaN   \n",
       "3                                    NaN   \n",
       "4                                    NaN   \n",
       "\n",
       "   Viral Load (lymphoid) copies/10e6 cells  \\\n",
       "0                                      NaN   \n",
       "1                                      NaN   \n",
       "2                                      NaN   \n",
       "3                                      NaN   \n",
       "4                                      NaN   \n",
       "\n",
       "                                       Tissue Source  Year of Death  \n",
       "0                                                NaN            NaN  \n",
       "1  University of California at San Francisco AIDS...            NaN  \n",
       "2  University of California at San Francisco AIDS...            NaN  \n",
       "3  University of California at San Francisco AIDS...            NaN  \n",
       "4                                                NaN            NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The metadata associated witht he Holman Training dataset.\n",
    "HolmanTrain = pd.read_csv('./HolmanAndGabuzda/HolmanTrain.csv')\n",
    "print(HolmanTrain.shape)\n",
    "HolmanTrain.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2552, 26)\n",
      "['Sample Tissue Code', 'Sample Tissue Class', 'Nucleic Acid Type', 'Sample Country', 'Sample City', 'Patient Age', 'Heath Status', 'Drug Naive', 'Antiretroviral Treatment', 'Viral Load Plasma', 'Viral Load Brain', 'Viral Load Lymphoid', 'CD4 Count', 'Neuropathological diagnosis', 'Neurocognitive diagnosis', 'Patient Code', 'Patient Sex', 'Patient Risk Factor', 'Patient Year of Death', 'Publication', 'PubMed Id', 'Publication Title', 'Sequence Start', 'Sequence Stop', 'Accession', 'Unnamed: 26']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ryanvelazquez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: from_csv is deprecated. Please use read_csv(...) instead. Note that some of the default arguments are different, so please refer to the documentation for from_csv when changing your function calls\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Tissue Code</th>\n",
       "      <th>Sample Tissue Class</th>\n",
       "      <th>Nucleic Acid Type</th>\n",
       "      <th>Sample Country</th>\n",
       "      <th>Sample City</th>\n",
       "      <th>Patient Age</th>\n",
       "      <th>Heath Status</th>\n",
       "      <th>Drug Naive</th>\n",
       "      <th>Antiretroviral Treatment</th>\n",
       "      <th>Viral Load Plasma</th>\n",
       "      <th>...</th>\n",
       "      <th>Patient Sex</th>\n",
       "      <th>Patient Risk Factor</th>\n",
       "      <th>Patient Year of Death</th>\n",
       "      <th>Publication</th>\n",
       "      <th>PubMed Id</th>\n",
       "      <th>Publication Title</th>\n",
       "      <th>Sequence Start</th>\n",
       "      <th>Sequence Stop</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Unnamed: 26</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample Tissue Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Spleen</th>\n",
       "      <td>FMA:7196</td>\n",
       "      <td>Blood and Lymphoid</td>\n",
       "      <td>DNA</td>\n",
       "      <td>US</td>\n",
       "      <td>Chicago</td>\n",
       "      <td>39</td>\n",
       "      <td>AIDS (deceased)</td>\n",
       "      <td>no</td>\n",
       "      <td>AZT</td>\n",
       "      <td>194190.0</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>male homosexual</td>\n",
       "      <td>1993.0</td>\n",
       "      <td>Thomas, 2007</td>\n",
       "      <td>17084877</td>\n",
       "      <td>Macrophage entry mediated by HIV Envs from bra...</td>\n",
       "      <td>6225</td>\n",
       "      <td>8795</td>\n",
       "      <td>DQ976393</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Spleen</th>\n",
       "      <td>FMA:7196</td>\n",
       "      <td>Blood and Lymphoid</td>\n",
       "      <td>DNA</td>\n",
       "      <td>US</td>\n",
       "      <td>Chicago</td>\n",
       "      <td>39</td>\n",
       "      <td>AIDS (deceased)</td>\n",
       "      <td>no</td>\n",
       "      <td>AZT</td>\n",
       "      <td>194190.0</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>male homosexual</td>\n",
       "      <td>1993.0</td>\n",
       "      <td>Thomas, 2007</td>\n",
       "      <td>17084877</td>\n",
       "      <td>Macrophage entry mediated by HIV Envs from bra...</td>\n",
       "      <td>6225</td>\n",
       "      <td>8795</td>\n",
       "      <td>DQ976392</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Spleen</th>\n",
       "      <td>FMA:7196</td>\n",
       "      <td>Blood and Lymphoid</td>\n",
       "      <td>DNA</td>\n",
       "      <td>US</td>\n",
       "      <td>Chicago</td>\n",
       "      <td>39</td>\n",
       "      <td>AIDS (deceased)</td>\n",
       "      <td>no</td>\n",
       "      <td>AZT</td>\n",
       "      <td>194190.0</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>male homosexual</td>\n",
       "      <td>1993.0</td>\n",
       "      <td>Thomas, 2007</td>\n",
       "      <td>17084877</td>\n",
       "      <td>Macrophage entry mediated by HIV Envs from bra...</td>\n",
       "      <td>6225</td>\n",
       "      <td>8795</td>\n",
       "      <td>DQ976391</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lymph node</th>\n",
       "      <td>FMA:5034</td>\n",
       "      <td>Blood and Lymphoid</td>\n",
       "      <td>DNA</td>\n",
       "      <td>US</td>\n",
       "      <td>Chicago</td>\n",
       "      <td>39</td>\n",
       "      <td>AIDS (deceased)</td>\n",
       "      <td>no</td>\n",
       "      <td>AZT</td>\n",
       "      <td>194190.0</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>male homosexual</td>\n",
       "      <td>1993.0</td>\n",
       "      <td>Thomas, 2007</td>\n",
       "      <td>17084877</td>\n",
       "      <td>Macrophage entry mediated by HIV Envs from bra...</td>\n",
       "      <td>6225</td>\n",
       "      <td>8795</td>\n",
       "      <td>DQ976390</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lymph node</th>\n",
       "      <td>FMA:5034</td>\n",
       "      <td>Blood and Lymphoid</td>\n",
       "      <td>DNA</td>\n",
       "      <td>US</td>\n",
       "      <td>Chicago</td>\n",
       "      <td>39</td>\n",
       "      <td>AIDS (deceased)</td>\n",
       "      <td>no</td>\n",
       "      <td>AZT</td>\n",
       "      <td>194190.0</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>male homosexual</td>\n",
       "      <td>1993.0</td>\n",
       "      <td>Thomas, 2007</td>\n",
       "      <td>17084877</td>\n",
       "      <td>Macrophage entry mediated by HIV Envs from bra...</td>\n",
       "      <td>6225</td>\n",
       "      <td>8795</td>\n",
       "      <td>DQ976389</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Sample Tissue Code Sample Tissue Class Nucleic Acid Type  \\\n",
       "Sample Tissue Name                                                            \n",
       "Spleen                       FMA:7196  Blood and Lymphoid               DNA   \n",
       "Spleen                       FMA:7196  Blood and Lymphoid               DNA   \n",
       "Spleen                       FMA:7196  Blood and Lymphoid               DNA   \n",
       "Lymph node                   FMA:5034  Blood and Lymphoid               DNA   \n",
       "Lymph node                   FMA:5034  Blood and Lymphoid               DNA   \n",
       "\n",
       "                   Sample Country Sample City Patient Age     Heath Status  \\\n",
       "Sample Tissue Name                                                           \n",
       "Spleen                         US     Chicago          39  AIDS (deceased)   \n",
       "Spleen                         US     Chicago          39  AIDS (deceased)   \n",
       "Spleen                         US     Chicago          39  AIDS (deceased)   \n",
       "Lymph node                     US     Chicago          39  AIDS (deceased)   \n",
       "Lymph node                     US     Chicago          39  AIDS (deceased)   \n",
       "\n",
       "                   Drug Naive Antiretroviral Treatment  Viral Load Plasma  \\\n",
       "Sample Tissue Name                                                          \n",
       "Spleen                     no                      AZT           194190.0   \n",
       "Spleen                     no                      AZT           194190.0   \n",
       "Spleen                     no                      AZT           194190.0   \n",
       "Lymph node                 no                      AZT           194190.0   \n",
       "Lymph node                 no                      AZT           194190.0   \n",
       "\n",
       "                       ...      Patient Sex  Patient Risk Factor  \\\n",
       "Sample Tissue Name     ...                                         \n",
       "Spleen                 ...                M      male homosexual   \n",
       "Spleen                 ...                M      male homosexual   \n",
       "Spleen                 ...                M      male homosexual   \n",
       "Lymph node             ...                M      male homosexual   \n",
       "Lymph node             ...                M      male homosexual   \n",
       "\n",
       "                    Patient Year of Death   Publication PubMed Id  \\\n",
       "Sample Tissue Name                                                  \n",
       "Spleen                             1993.0  Thomas, 2007  17084877   \n",
       "Spleen                             1993.0  Thomas, 2007  17084877   \n",
       "Spleen                             1993.0  Thomas, 2007  17084877   \n",
       "Lymph node                         1993.0  Thomas, 2007  17084877   \n",
       "Lymph node                         1993.0  Thomas, 2007  17084877   \n",
       "\n",
       "                                                    Publication Title  \\\n",
       "Sample Tissue Name                                                      \n",
       "Spleen              Macrophage entry mediated by HIV Envs from bra...   \n",
       "Spleen              Macrophage entry mediated by HIV Envs from bra...   \n",
       "Spleen              Macrophage entry mediated by HIV Envs from bra...   \n",
       "Lymph node          Macrophage entry mediated by HIV Envs from bra...   \n",
       "Lymph node          Macrophage entry mediated by HIV Envs from bra...   \n",
       "\n",
       "                   Sequence Start Sequence Stop  Accession Unnamed: 26  \n",
       "Sample Tissue Name                                                      \n",
       "Spleen                       6225          8795   DQ976393         NaN  \n",
       "Spleen                       6225          8795   DQ976392         NaN  \n",
       "Spleen                       6225          8795   DQ976391         NaN  \n",
       "Lymph node                   6225          8795   DQ976390         NaN  \n",
       "Lymph node                   6225          8795   DQ976389         NaN  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# HIV Brain Sequence Database tsv file downloaded on 10-10-2018\n",
    "HIVSeqDBMetadata = pd.DataFrame.from_csv('./HIVBrainSeqDB/HIVSeqDB-MetaData.tsv', sep='\\t')\n",
    "print(HIVSeqDBMetadata.shape)\n",
    "print(list(HIVSeqDBMetadata.columns))\n",
    "HIVSeqDBMetadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Some Preliminary Questions About the Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. Why are there more rows in the metadata.csv then there are sequences (5,386 vs 2,494)... the paper says 2,492 sequences?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are  5386  unique sequences in the metadata.csv file\n",
      "There are  2494  sequences in the Metadataset fasta file\n",
      "There are  2494 sequences in the fasta that have a matching row in the csv\n"
     ]
    }
   ],
   "source": [
    "print('There are ',len(metadata.Accession.unique()), ' unique sequences in the metadata.csv file')\n",
    "print('There are ',len(Files['C2V3C3_Metadataset_AA.fasta'][\"sequences\"]), ' sequences in the Metadataset fasta file')\n",
    "\n",
    "# I'm not sure of the reason for this but we can easily widdle down the metadata.csv file to only the sequences in the fasta file.\n",
    "numberOfSequencesInTheFastaWithARowInTheCSV = 0\n",
    "for index, row in metadata.iterrows():\n",
    "    if row['Accession'] in Files['C2V3C3_Metadataset_AA.fasta'][\"ids\"]:\n",
    "        numberOfSequencesInTheFastaWithARowInTheCSV +=1\n",
    "print('There are ', numberOfSequencesInTheFastaWithARowInTheCSV, 'sequences in the fasta that have a matching row in the csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So... \n",
    "- There are over twice as many rows in the metadata.csv file than in the metadata.fasta file.\n",
    "- All the sequences in the metadata.fasta file have a corresponding row in the csv file\n",
    "- I'm not sure what the extra rows are..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. How many sequences are shared between files?\n",
    "We won't look at the LANL file for now because we'll assume many of the sequences in other files are in there."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of shared sequences\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LANLOgishiAA</th>\n",
       "      <th>metaOgishiAA</th>\n",
       "      <th>metaOgishi</th>\n",
       "      <th>HANDdbOgishi</th>\n",
       "      <th>BraindbHepler</th>\n",
       "      <th>BrainDB</th>\n",
       "      <th>valHolman</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LANLOgishiAA</th>\n",
       "      <td>21609</td>\n",
       "      <td>1197</td>\n",
       "      <td>1568</td>\n",
       "      <td>697</td>\n",
       "      <td>217</td>\n",
       "      <td>424</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metaOgishiAA</th>\n",
       "      <td>1197</td>\n",
       "      <td>2494</td>\n",
       "      <td>2494</td>\n",
       "      <td>1449</td>\n",
       "      <td>503</td>\n",
       "      <td>915</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metaOgishi</th>\n",
       "      <td>1568</td>\n",
       "      <td>2494</td>\n",
       "      <td>5386</td>\n",
       "      <td>1449</td>\n",
       "      <td>939</td>\n",
       "      <td>1917</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HANDdbOgishi</th>\n",
       "      <td>697</td>\n",
       "      <td>1449</td>\n",
       "      <td>1449</td>\n",
       "      <td>1690</td>\n",
       "      <td>483</td>\n",
       "      <td>912</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BraindbHepler</th>\n",
       "      <td>217</td>\n",
       "      <td>503</td>\n",
       "      <td>939</td>\n",
       "      <td>483</td>\n",
       "      <td>1085</td>\n",
       "      <td>1085</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BrainDB</th>\n",
       "      <td>424</td>\n",
       "      <td>915</td>\n",
       "      <td>1917</td>\n",
       "      <td>912</td>\n",
       "      <td>1085</td>\n",
       "      <td>2552</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>valHolman</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               LANLOgishiAA  metaOgishiAA  metaOgishi  HANDdbOgishi  \\\n",
       "LANLOgishiAA          21609          1197        1568           697   \n",
       "metaOgishiAA           1197          2494        2494          1449   \n",
       "metaOgishi             1568          2494        5386          1449   \n",
       "HANDdbOgishi            697          1449        1449          1690   \n",
       "BraindbHepler           217           503         939           483   \n",
       "BrainDB                 424           915        1917           912   \n",
       "valHolman                 0             0           0             0   \n",
       "\n",
       "               BraindbHepler  BrainDB  valHolman  \n",
       "LANLOgishiAA             217      424          0  \n",
       "metaOgishiAA             503      915          0  \n",
       "metaOgishi               939     1917          0  \n",
       "HANDdbOgishi             483      912          0  \n",
       "BraindbHepler           1085     1085          0  \n",
       "BrainDB                 1085     2552          0  \n",
       "valHolman                  0        0         75  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assign the lists of sequences to some short variables for easy access.\n",
    "seqs={\n",
    "    'LANLOgishiAA': Files['C2V3C3_LANL_AA.fasta']['ids'],\n",
    "    'metaOgishiAA': Files['C2V3C3_Metadataset_AA.fasta']['ids'],\n",
    "    'metaOgishi': list(metadata['Accession']),\n",
    "    'HANDdbOgishi': Files['C2V3C3_HANDDatabase_AA.fasta']['ids'],\n",
    "    'BraindbHepler': Files['HIVBrainSeqDBHepler.fasta']['ids'],\n",
    "    'BrainDB': Files['HIVBrainSeqDB.fasta']['ids'],\n",
    "    'valHolman': Files['HAD.fas']['ids']\n",
    "}\n",
    "# Create a matrix of overlaps\n",
    "seqNumberMatrix = []\n",
    "for key in seqs:\n",
    "    LANLOverlap = len(set(seqs[key]).intersection(seqs['LANLOgishiAA']))\n",
    "    metaOverlapAA = len(set(seqs[key]).intersection(seqs['metaOgishiAA']))\n",
    "    metaOverlap = len(set(seqs[key]).intersection(seqs['metaOgishi']))\n",
    "    HANDdbOgishiOverlap = len(set(seqs[key]).intersection(seqs['HANDdbOgishi']))\n",
    "    BraindbHeplerOverlap = len(set(seqs[key]).intersection(seqs['BraindbHepler']))\n",
    "    BrainDBOverlap = len(set(seqs[key]).intersection(seqs['BrainDB']))\n",
    "    valHolmanOverlap = len(set(seqs[key]).intersection(seqs['valHolman']))\n",
    "    \n",
    "    seqNumberMatrix.append([LANLOverlap, metaOverlapAA, metaOverlap, HANDdbOgishiOverlap, BraindbHeplerOverlap, BrainDBOverlap, valHolmanOverlap])\n",
    "# Create and print pandas DF\n",
    "sharedSeqs = pd.DataFrame(seqNumberMatrix, columns = list(seqs.keys()), index = list(seqs.keys()))\n",
    "print('Number of shared sequences')\n",
    "sharedSeqs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### c. How many sequences are there in total?\n",
    "(excluding LANL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "excluding sequences from Ogishi without AA data\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "4352"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('excluding sequences from Ogishi without AA data')\n",
    "len(set(seqs['metaOgishiAA'] + seqs['HANDdbOgishi'] + seqs['BraindbHepler'] + seqs['BrainDB'] + seqs['valHolman']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "including sequences from Ogishi without AA data\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "6242"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('including sequences from Ogishi without AA data')\n",
    "len(set(seqs['metaOgishi'] + seqs['metaOgishiAA'] + seqs['HANDdbOgishi'] + seqs['BraindbHepler'] + seqs['BrainDB'] + seqs['valHolman']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### d. How many patients are there in total?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This question (as well as others, including number 4 above) are probably best answered by trying to assemble a complete dataset with associated metadata. See the \"Munge the Data\" notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pobably some where between  183  and  211  patients.\n"
     ]
    }
   ],
   "source": [
    "# I started taking a look at this with:\n",
    "patientIds = list(metadata['LANL_Patient.Id'])\n",
    "patientIdsNoNAN = list(filter(lambda x: x >= 0, set(patientIds)))\n",
    "print('pobably some where between ', len(set(patientIdsNoNAN)), ' and ',len(set(patientIds)), ' patients.')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
